src/error.rs

```diff
@@ -0,0 +1,11 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum IcebergError {
```

Comment:
I perfer to have a custom error that we can have the following things:

- Error code that help users to decide how to take actions.
- Error Message that describe what happend.
- Error Context that provide more information for users.
- Error Causes that help users to figure out what happened (with backtrace support).

I'm willing to implement this after this PR.

Reply:
Sorry for jumping ahead and use thiserror, it looked really convenient and powerful at the same time. But we can also use a custom error. Just for reference I think most of the functionality(error message, source, backtrace) can be implemented with thiserror plus it's really convenient to impl `From` for external Errors.

Reply:
Wrapping all errors with `thiserror` is convenient for developers but not good for users. My primary concern is how users will handle the returning errors.

For example, does it meaningful for users to distinguish between `utf8 error`, `json format error` or `avro error`? By providing the custom error by us, users can decide which action to take by error kind and know more context while needed.

Reply:
It would leave it for now as is and if it's okay for you, you can implement the custom error later.